firefox-ios/Client/Frontend/Browser/Tabs/Views/TabAnimation.swift

```diff
@@ -0,0 +1,286 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import UIKit
+import Shared
+
+extension TabTrayViewController: UIViewControllerTransitioningDelegate {
+    func animationController(
+        forPresented presented: UIViewController,
+        presenting: UIViewController,
+        source: UIViewController
+    ) -> UIViewControllerAnimatedTransitioning? {
+        return BasicAnimationController(delegate: self, direction: .presenting)
+    }
+
+    func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
+        return BasicAnimationController(delegate: self, direction: .dismissing)
+    }
+}
+
+extension TabTrayViewController: BasicAnimationControllerDelegate {
+    func animatePresentation(context: UIViewControllerContextTransitioning) {
+        guard
+            let containerController = context.viewController(forKey: .from) as? UINavigationController,
+            let bvc = containerController.topViewController as? BrowserViewController,
+            let destinationController = context.viewController(forKey: .to)
+        else {
+            logger.log(
+        """
+            Attempted to present the tab tray on something that is not a BrowserViewController which is
+            currently unsupported.
+        """,
+        level: .warning,
+        category: .tabs
+            )
+            context.completeTransition(true)
+            return
+        }
+
+        guard let selectedTab = bvc.tabManager.selectedTab,
+              let webView = selectedTab.webView
+        else {
+            logger.log("Attempted to present the tab tray without having a selected tab",
+                       level: .warning,
+                       category: .tabs)
+            context.completeTransition(true)
+            return
+        }
+
+        let finalFrame = context.finalFrame(for: destinationController)
+
+        // Tab snapshot animates from web view container on BVC to the cell frame
+        let tabSnapshot = UIImageView(image: selectedTab.screenshot ?? .init())
+        tabSnapshot.layer.cornerCurve = .continuous
+        tabSnapshot.clipsToBounds = true
+        tabSnapshot.contentMode = .scaleAspectFill
+        tabSnapshot.frame = webView.frame
+
+        // Allow the UI to render to make the snapshotting code more performant
+        // swiftlint:disable closure_body_length
```

Comment:
To avoid the `closure_body_length` could we have this code under another function? If it's not too complicated to change, otherwise you can disregard this comment 👍